src/Processors/QueryPlan/ReadFromMergeTree.cpp

```diff
@@ -1843,13 +1843,34 @@ static void buildIndexes(
             skip_indexes.useful_indices.emplace_back(index_helper, condition);
     }
 
+    std::unordered_map<String, UInt64> max_index_sizes;
+   
+    for (auto idx : skip_indexes.useful_indices)
+    {
+        UInt64 max_size = 0;
+        for (auto &part : parts)
+        {
+            auto extension = idx.index->getDeserializedFormat(part.data_part->getDataPartStorage(), idx.index->getFileName()).extension;
+            auto sz = part.data_part->getFileSizeOrZero(idx.index->getFileName() + extension);
+            max_size = sz > max_size ? sz : max_size;
+        }
+        max_index_sizes[idx.index->getFileName()] = max_size;
+    }
+
     // Move minmax indices to first positions, so they will be applied first as cheapest ones
-    std::stable_sort(skip_indexes.useful_indices.begin(), skip_indexes.useful_indices.end(), [](const auto & l, const auto & r)
+    std::stable_sort(skip_indexes.useful_indices.begin(), skip_indexes.useful_indices.end(), [ &max_index_sizes](const auto & l, const auto & r)
     {
         bool l_is_minmax = typeid_cast<const MergeTreeIndexMinMax *>(l.index.get());
         bool r_is_minmax = typeid_cast<const MergeTreeIndexMinMax *>(r.index.get());
         if (l_is_minmax == r_is_minmax)
-            return false;
+        {
+            const auto l_granularity = l.index->getGranularity();
+            const auto r_granularity = r.index->getGranularity();
+
+            const auto l_size = max_index_sizes[l.index->getFileName()];
+            const auto r_size = max_index_sizes[r.index->getFileName()];
+            return (l_size < r_size) || (l_granularity > r_granularity);
```

Comment:
For minmax indexes these are equivalent. Not immediately clear to me what the order of operands should be for the general case. I'm guessing index files are small enough that we are typically not bound to file download (in SharedMergeTree)/reading, and instead bound to number of index evaluations (which scales with granularity). i.e. maybe granularity should be the first operand still?